### PR TITLE
/cmd followups

### PR DESCRIPTION
**Pull request by mordamax**
Closes: https://github.com/paritytech/polkadot-sdk/issues/5545

- add missing template for frame & xcm benchmarks
- fix `git pull` -> https://github.com/paritytech/polkadot-sdk/actions/runs/10644887539/job/29510118915
- respect runtimes headers - use GNU instead of apache for runtimes
- adds tests for cmd.py

Tip: review this one with Whitespace hidden
image

### CODE REVIEW

**Comment by mordamax** [comment]
/cmd bench --runtime westend --pallet pallet_balances --clean

**Comment by mordamax** [comment]
/cmd bench --pallet pallet_balances --continue-on-fail --clean

**Comment by mordamax** [comment]
/cmd bench --pallet pallet_balances --continue-on-fail --clean

**Comment by mordamax** [comment]
/cmd bench --pallet pallet_balances --continue-on-fail --clean

**Comment by mordamax** [comment]
/cmd bench --runtime asset-hub-westend --pallet pallet_balances --continue-on-fail --clean

**Review by EgorPopelyaev — APPROVED** [review]

**Review by alvicsam — APPROVED** [review]

File: .github/scripts/cmd/cmd.py [hunk 1]
```diff
@@ -67,130 +69,153 @@
...
+            for runtime, pallets in failed_benchmarks.items():
+                print(f'-- {runtime}: {pallets}')
+
+        if successful_benchmarks:
+            print('✅ Successful benchmarks of runtimes/pallets:')
+            for runtime, pallets in successful_benchmarks.items():
+                print(f'-- {runtime}: {pallets}')
+
+    elif args.command == 'fmt':
+        command = f"cargo +nightly fmt"
```
Comment (muharem) [inline .github/scripts/cmd/cmd.py:200]: we should make sure we use the same version as in CI job checking the formatting

Comment (mordamax) [reply .github/scripts/cmd/cmd.py:200]: 👍 true, it's the same, as the /cmd uses ci-unified image, where the similar rust / nightly versions are installed

**Review by muharem — APPROVED** [review]
new to this code, so please take this approval with that in mind.

**Comment by muharem** [comment]
maybe worth having @ggwpez eyes on this

**Review by ggwpez — APPROVED** [review]

**Comment by s0me0ne-unkn0wn** [comment]
@mordamax 

How did it happen that this PR changed benchmarked weights for the `balances` pallet so much? Is this an expected outcome?

From what I see in the output generated by the benchmark, I'm not even sure it was run on the reference hardware.

**Comment by mordamax** [comment]
@s0me0ne-unkn0wn 

1.as for reference or not - it was generated using a new github runner (vs gitlab runner) - `parity-weights` - check its specs here https://github.com/paritytech/ci_cd/wiki/GitHub
2. [last update](https://github.com/paritytech/polkadot-sdk/commit/c3e57c1b3b9fdd768a57e91d353d57f6bbff7ce5) was on May 8, so not sure what exactly might have impacted in between of this vs that

And to test old vs new I think we can run the same pallet somewhere via old and new command bot 

`/cmd bench --runtime westend --pallet pallet_balances`
or 
`bot bench polkadot-pallet --pallet=pallet_balances`

Let's see here https://github.com/paritytech/polkadot-sdk/pull/6095

**Comment by mordamax** [comment]
@s0me0ne-unkn0wn you're right, looks like there's a consistent and significant difference, thanks for flagging!

frame-omni-bencher + github runner (parity-weights) -> https://github.com/paritytech/polkadot-sdk/pull/6095/commits/cbb16b4ad383cc9c060fcdc6c67288a2a0ffe3ae
benchmark + gitlab runner (weights) -> https://github.com/paritytech/polkadot-sdk/pull/6095/commits/2eeba65b172056a499dd196416f5f6f69846c04c

@alvicsam could you please check and share more details what could lead to that consistent difference between the 2 runners?
@ggwpez do you know if the switch to frame-omni-bencher could potentially be part of a reason?

**Comment by alvicsam** [comment]
The hardware is the same: `n2-standard-8` with option `threads_per_core=1` and local ssd. The runners environment differs (VM + docker in case of gitlab and k8s + docker-in-docker in case of github) but IMO it would be nice to test github runner with old benchmark to check the result first.

**Comment by ggwpez** [comment]
We could also add run `polkadot benchmark machine` at the beginning of the jobs to check the performance.

**Comment by mordamax** [comment]
GitHub Selfhosted runner VM
image

GitLab runner VM
image

but still the gitlab returns with more "better" around ~40-50% results

**Comment by s0me0ne-unkn0wn** [comment]
> we think that it might worth just recalculating the new baselines with new VMs

While I don't have better ideas myself, I'm still not sure it's the best solution. The first reason is the sTPS benchmark. Having 40% more reference time per transaction means we pack 40% fewer transactions into a block and thus we have 40% less TPS. Technically it doesn't make much difference, but from the marketing perspective 40% fall of TPS value of the whole network out of the blue is something that can make a lot of people unhappy.

The second reason is our attitude towards the end users, namely, validator operators. Like, we guarantee that someone having the given hardware spec can run a validator, and that's why we measure reference times with reference hardware. Increasing reference times by 40% means we either were undershooting before, which is terrible, or are overshooting now, which is not great as well. Our end-user-facing metrics should be as consistent as possible.

Last but not least, I'm just curious what the hell is going on there ;) I understand that different VMs having the same amount of resources will behave differently from the performance perspective, but a 40% difference is well over the threshold of something that can be explained solely by VM architecture differences, given that they both have very similar machine benchmarks.

That said, I feel like more justification is needed to re-run benchmarks and increase all the weights.

CC @eskimor

**Comment by ggwpez** [comment]
You are using what runtime for sTPS testing that @s0me0ne-unkn0wn? The Polkadot and Kusama runtimes have their own weights in the runtimes repo. Any weights we do here are for testnets only. Nobody should use them in prod, but generate their own anyway.

But I think if this Docker-in-docker-in-kubernets is 40% slower than an actual server, then we should not use it. Nobody would reasonably run a validator like this either, so its not really "reference hardware" anymore. Seems like the machine benchmarks dont really show this.

**Comment by s0me0ne-unkn0wn** [comment]
> What runtime are you using for sTPS testing

We've always used Rococo runtime for that and consciously used the Substrate weights without re-weighting. Substrate weights are the reference. I mean, if your runtime has `u128` balances and `AccountId32` accounts, etc., there should be no chance of getting *very* different measurements for your runtime IIUC.

Besides that, while relay chain runtimes, indeed, have their own weights, parachains often do not. And even if they had, it wouldn't help that much in measuring parachain throughput to have a hundred different weights for different parachains. So, for parachain throughput measurements, we also use Substrate weights as a reference.

> Nobody would reasonably run a validator like this either, so its not really "reference hardware" anymore.

Exactly! That's the point. Reference should be, well, a reference, that is, something we don't change easily.

**Comment by ggwpez** [comment]
>We've always used Rococo runtime for that and consciously used the Substrate weights without re-weighting. Substrate weights are the reference. I mean, if your runtime has u128 balances and AccountId32 accounts, etc., there should be no chance of getting very different measurements for your runtime IIUC.

If you look only at the balances pallet, then I guess, yes. Generally though, no. The weight of each pallet depends on the runtime config. The callbacks that are configured and the general pallet config can cause hugely different weights for production cases.

The weights in the runtimes repo are more reliable, if you want to do some testing. Either way, for getting the maximal number of TPS, the weights are not really important. You should just increase the block weight limit until the relay block time increases or the PVF times out.  
The weights are always an over-estimate, so if you want maximal numbers then you need to counter-correct for that.

**Comment by s0me0ne-unkn0wn** [comment]
> The weights are always an over-estimate, so if you want maximal numbers then you need to counter-correct for that.

I understand that, but we don't want maximum numbers but fair numbers. Like, if someone asks, "If I start a parachain on Polkadot, what TPS will I get?" and we answer, "You'll get 5550 balance transfers per block given that you're not using weird data types in your setup", that's fair. Yes, we could maximize even that (like, parameterize the balance transfer benchmark and account for the shortest execution path where we transfer from a pre-funded to a pre-existing account and thus get a lot more TPS in the sTPS test), but that wouldn't be *that* fair anymore. So, I believe we're just trying to get the right balance point between being precise and being fair.

**Pull request merged**